### PR TITLE
Delete Shape Refactor

### PR DESCRIPTION
**Pull request by br648**
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [x] The description lists all applicable issues this PR seeks to resolve
- [x] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

The recent update to delete shapes if a route or pattern is deleted does not take into consideration that a shape may still be in use. This refactor checks if a shape is being used by more than one trip/pattern/route. If this is the case, it is not deleted.

Jira: https://ibisolutions.atlassian.net/browse/DT-143

### CODE REVIEW

**Review by miles-grant-ibigroup — APPROVED** [review]
Looks clean and I've run locally and deletion seems to work great! Thanks for the changes

File: src/test/java/com/conveyal/gtfs/loader/JDBCTableWriterTest.java [hunk 1]
```diff
@@ -549,14 +549,59 @@ public void shouldDeleteReferencingTripsAndStopTimesOnPatternDelete() throws IOE
     }
 
     /**
-     * Deleting a route should cascade delete referencing patterns, trips and stop times.
+     * Deleting a route should cascade delete referencing patterns, trips, stop times and shapes.
```
Comment (binh-dam-ibigroup) [inline src/test/java/com/conveyal/gtfs/loader/JDBCTableWriterTest.java]: This comment seems now related to method `createThenDeleteRoute`.

File: src/test/java/com/conveyal/gtfs/loader/JDBCTableWriterTest.java [hunk 2]
```diff
@@ -549,14 +549,59 @@ public void shouldDeleteReferencingTripsAndStopTimesOnPatternDelete() throws IOE
...
+     * Deleting a route using a shared shape, should cascade delete referencing patterns, trips and stop times, but not
+     * shapes.
+     */
+    @Test
+    void shouldKeepShapeOnCascadeDeleteOfRoute() throws IOException, SQLException, InvalidNamespaceException {
+        String routeId = "8472017";
+        String shapeId = "sharedShapeId";
+
+        // Create a pattern which uses the same shape as the pattern that will be deleted. This is to prevent the shape
+        // from being deleted.
```
Comment (binh-dam-ibigroup) [inline src/test/java/com/conveyal/gtfs/loader/JDBCTableWriterTest.java:579]: Remind in this comment where the number 4 comes from.

File: src/main/java/com/conveyal/gtfs/model/Shape.java [hunk 3]
```diff
@@ -28,4 +40,126 @@ public Shape (GTFSFeed feed, String shape_id) {
...
+        String shapesTable,
+        String routeOrPatternId) throws SQLException {
+
+        String sql = (routeOrPatternIdColumn.equals("pattern_id"))
+            ? String.format(
+            "select shape_id from %s s join %s p using (shape_id) where s.shape_id = p.shape_id and p.pattern_id = '%s'",
+            shapesTable,
+            patternsTable,
+            routeOrPatternId
+        )
```
Comment (binh-dam-ibigroup) [inline src/main/java/com/conveyal/gtfs/model/Shape.java]: Indent these lines (+1 tab).

File: src/main/java/com/conveyal/gtfs/model/Shape.java [hunk 4]
```diff
@@ -28,4 +40,126 @@ public Shape (GTFSFeed feed, String shape_id) {
...
+            : String.format(
+            "select s.shape_id from %s s inner join %s p on s.shape_id = p.shape_id inner join %s r on p.route_id = r.route_id and r.route_id = '%s'",
+            shapesTable,
+            patternsTable,
+            String.format("%s.routes", tablePrefix),
+            routeOrPatternId
+        );
```
Comment (binh-dam-ibigroup) [inline src/main/java/com/conveyal/gtfs/model/Shape.java]: Add +1 tab.

File: src/main/java/com/conveyal/gtfs/model/Shape.java [hunk 5]
```diff
@@ -28,4 +40,126 @@ public Shape (GTFSFeed feed, String shape_id) {
...
+        }
+
+        if (!shapeIdsToDelete.isEmpty()) {
+            String sql = String.format(
+                "delete from %s where shape_id in (%s)",
+                shapesTable,
+                StringUtils.wrap(String.join(",", shapeIdsToDelete), "'")
+            );
+            try (Statement statement = connection.createStatement()) {
+                LOG.info("{}", sql);
```
Comment (binh-dam-ibigroup) [inline src/main/java/com/conveyal/gtfs/model/Shape.java]: Remove first argument.

File: src/main/java/com/conveyal/gtfs/model/Shape.java [hunk 6]
```diff
@@ -28,4 +40,126 @@ public Shape (GTFSFeed feed, String shape_id) {
...
+
+        return getListOfShapeIds(connection, sql);
+    }
+
+    /**
+     * Extract and return a list of shape ids from the provided sql.
+     */
+    private static Set<String> getListOfShapeIds(Connection connection, String sql) throws SQLException {
+        PreparedStatement statement = connection.prepareStatement(sql);
+        LOG.info("{}", statement);
```
Comment (binh-dam-ibigroup) [inline src/main/java/com/conveyal/gtfs/model/Shape.java]: Remove first argument. Do we need a log statement here?

**Review by binh-dam-ibigroup — APPROVED** [review]
LGTM. See few minor comments.

**Comment by binh-dam-ibigroup** [comment]
> LGTM. See few minor comments.

Actually @br648 Could you add a `fix` commit?

**Pull request merged**